In [1]:
!pwd

/root/autodl-tmp/DeepSeek-OCR-main/DeepSeek-OCR-master


# 图片识别

## Prompt
```python
PROMPT = '<image>\n<|grounding|>Convert the document to markdown.'
# PROMPT = '<image>\nFree OCR.'
# TODO commonly used prompts
# document: <image>\n<|grounding|>Convert the document to markdown.
# other image: <image>\n<|grounding|>OCR this image.
# without layouts: <image>\nFree OCR.
# figures in document: <image>\nParse the figure.
# general: <image>\nDescribe this image in detail.
# rec: <image>\nLocate <|ref|>xxxx<|/ref|> in the image.
```


![测试图片](input/img/deepseekOCR.png)

## 1. 图片描述
- `prompt = "<image>\nDescribe this image in detail."`

In [4]:
from transformers import AutoModel, AutoTokenizer
import torch
import os
os.environ["CUDA_VISIBLE_DEVICES"] = '0'
model_name = 'deepseek-ai/DeepSeek-OCR'

tokenizer = AutoTokenizer.from_pretrained(model_name, trust_remote_code=True)
model = AutoModel.from_pretrained(model_name, _attn_implementation='flash_attention_2', trust_remote_code=True, use_safetensors=True)
model = model.eval().cuda().to(torch.bfloat16)

prompt = "<image>\nDescribe this image in detail."
# prompt = "<image>\nFree OCR. "
# prompt = "<image>\n<|grounding|>Convert the document to markdown. "
image_file = 'input/img/deepseekOCR.png'
output_path = 'img_out/Describe'

res = model.infer(tokenizer, prompt=prompt, image_file=image_file, output_path = output_path, base_size = 1024, image_size = 640, crop_mode=True, save_results = True, test_compress = True)


You are using a model of type deepseek_vl_v2 to instantiate a model of type DeepseekOCR. This is not supported for all configurations of models and can yield errors.
Some weights of DeepseekOCRForCausalLM were not initialized from the model checkpoint at deepseek-ai/DeepSeek-OCR and are newly initialized: ['model.vision_model.embeddings.position_ids']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/root/miniconda3/envs/deepseek-ocr/lib/python3.12/site-packages/transformers/generation/configuration_utils.py:590: UserWarning: `do_sample` is set to `False`. However, `temperature` is set to `0.0` -- this flag is only used in sample-based generation modes. You should set `do_sample=True` or unset `temperature`.
  warnings.warn(
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` t

BASE:  torch.Size([1, 256, 1280])
PATCHES:  torch.Size([8, 100, 1280])
The figure consists of two parts: (a) Compression on the Fox benchmark and (b) Performance on Omnidocbench.

### Part (a): Compression on Fox Benchmark

This part of the figure shows the compression ratio achieved by the proposed method for various text token sizes. The compression ratio is represented by the height of the bars, with each bar corresponding to a specific text token size. The bars are color-coded to indicate different compression ratios:

- **64 bits tokens (left)**: Represented by the lightest shade of purple.
- **100 bits tokens (middle)**: Represented by a slightly darker shade of purple.
- **64 bits tokens (right)**: Represented by a medium shade of purple.
- **100 bits tokens (right)**: Represented by the darkest shade of purple.

The x-axis represents the text tokens in per page (Ground-truth), ranging from 600 to 1200. The y-axis represents the precision percentage, ranging from 0% to 100%. Eac

image: 0it [00:00, ?it/s]
other: 0it [00:00, ?it/s]


## 2. Free OCR
- `prompt = "<image>\nFree OCR. "`

In [5]:
from transformers import AutoModel, AutoTokenizer
import torch
import os
os.environ["CUDA_VISIBLE_DEVICES"] = '0'
model_name = 'deepseek-ai/DeepSeek-OCR'

tokenizer = AutoTokenizer.from_pretrained(model_name, trust_remote_code=True)
model = AutoModel.from_pretrained(model_name, _attn_implementation='flash_attention_2', trust_remote_code=True, use_safetensors=True)
model = model.eval().cuda().to(torch.bfloat16)

# prompt = "<image>\nDescribe this image in detail."
prompt = "<image>\nFree OCR. "
# prompt = "<image>\n<|grounding|>Convert the document to markdown. "
image_file = 'input/img/deepseekOCR.png'
output_path = 'img_out/Free OCR'

res = model.infer(tokenizer, prompt=prompt, image_file=image_file, output_path = output_path, base_size = 1024, image_size = 640, crop_mode=True, save_results = True, test_compress = True)


You are using a model of type deepseek_vl_v2 to instantiate a model of type DeepseekOCR. This is not supported for all configurations of models and can yield errors.
Some weights of DeepseekOCRForCausalLM were not initialized from the model checkpoint at deepseek-ai/DeepSeek-OCR and are newly initialized: ['model.vision_model.embeddings.position_ids']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/root/miniconda3/envs/deepseek-ocr/lib/python3.12/site-packages/transformers/generation/configuration_utils.py:590: UserWarning: `do_sample` is set to `False`. However, `temperature` is set to `0.0` -- this flag is only used in sample-based generation modes. You should set `do_sample=True` or unset `temperature`.
  warnings.warn(
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` t

BASE:  torch.Size([1, 256, 1280])
PATCHES:  torch.Size([8, 100, 1280])
<table><tr><td>Model</td><td>Text Tokens in Per Page (Ground-truth)</td><td>Overall Performance (Edit Distance)</td><td>Performance on Omnipodbench</td></tr><tr><td>64 vis toks(left)</td><td>600-700</td><td>10.5</td><td>DeepSeek-OCR (Gundam-M 200dpi)</td></tr><tr><td>64 vis toks(left)</td><td>700-800</td><td>11.8</td><td>DeepSeek-OCR (Large)</td></tr><tr><td>64 vis toks(left)</td><td>800-900</td><td>13.2</td><td>DeepSeek-OCR (Small)</td></tr><tr><td>64 vis toks(left)</td><td>900-1000</td><td>15.1</td><td>DeepSeek-OCR (Tiny)</td></tr><tr><td>64 vis toks(left)</td><td>1000-1100</td><td>15.5</td><td>DeepSeek-OCR (Tiny)</td></tr><tr><td>64 vis toks(right)</td><td>1100-1200</td><td>11.3</td><td>DeepSeek-OCR (Tiny)</td></tr><tr><td>64 vis toks (right)</td><td>1200-1300</td><td>10.6</td><td>DeepSeek-OCR (Tiny)</td></tr><tr><td>100 vis toks(left)</td><td>600-700</td><td>10.5</td><td>DeepSeer-OCR (Gundam-M 200dpi)</td></tr><

image: 0it [00:00, ?it/s]
other: 0it [00:00, ?it/s]


<table><tr><td>Model</td><td>Text Tokens in Per Page (Ground-truth)</td><td>Overall Performance (Edit Distance)</td><td>Performance on Omnipodbench</td></tr><tr><td>64 vis toks(left)</td><td>600-700</td><td>10.5</td><td>DeepSeek-OCR (Gundam-M 200dpi)</td></tr><tr><td>64 vis toks(left)</td><td>700-800</td><td>11.8</td><td>DeepSeek-OCR (Large)</td></tr><tr><td>64 vis toks(left)</td><td>800-900</td><td>13.2</td><td>DeepSeek-OCR (Small)</td></tr><tr><td>64 vis toks(left)</td><td>900-1000</td><td>15.1</td><td>DeepSeek-OCR (Tiny)</td></tr><tr><td>64 vis toks(left)</td><td>1000-1100</td><td>15.5</td><td>DeepSeek-OCR (Tiny)</td></tr><tr><td>64 vis toks(right)</td><td>1100-1200</td><td>11.3</td><td>DeepSeek-OCR (Tiny)</td></tr><tr><td>64 vis toks (right)</td><td>1200-1300</td><td>10.6</td><td>DeepSeek-OCR (Tiny)</td></tr><tr><td>100 vis toks(left)</td><td>600-700</td><td>10.5</td><td>DeepSeer-OCR (Gundam-M 200dpi)</td></tr><tr><td>100 vis toks(left)</td><td>700-800</td><td>11.8</td><td>DeepSeer-OCR (Large)</td></tr><tr><td>100 vis toks(left)</td><td>800-900</td><td>13.2</td><td>DeepSeer-OCR (Small)</td></tr><tr><td>100 vis toks(left)</td><td>900-1000</td><td>15.1</td><td>Deep Seer-OCR (Tiny)</td></tr><tr><td>100 vis toks(left)</td><td>1000-1100</td><td>15.5</td><td>DeepsSeer-OCR (Tiny)</td></tr><tr><td>100 vis toks(left)</td><td>1100-1200</td><td>11.3</td><td>DeepSeer-OCR (Tiny)</td></tr><tr><td>100 vis toks(right)</td><td>1200-1300</td><td>10.6</td><td>DeepSeer-OCR (Tiny)</td></tr><tr><td>100 vis toks (right)</td><td>1300-1400</td><td>10.5</td><td>DeepSeer-OCR (Tiny)</td></tr><tr><td>100 vis toks</td><td>1400-1500</td><td>10.4</td><td>DeepSeer-OCR (Tiny)</td></tr><tr><td>100 vis toks (>1500)</td><td>1400-1500</td><td>10.3</td><td>DeepSeer-OCR (Tiny)</td></tr><tr><td>100 vis Tokens (>1500)</td><td>1500-1600</td><td>10.2</td><td>DeepSeer-OCR (Tiny)</td></tr><tr><td>100 vis Tokens (1500-1700)</td><td>1600-1700</td><td>10.1</td><td>DeepSeer-OCR (Tiny)</td></tr><tr><td>100 vis Tokens < 1500</td><td>1700-1800</td><td>10.0</td><td>DeepSeer-OCR (Tiny)</td></tr><tr><td>100 vis Tokens</td><td>1800-1900</td><td>10.0</td><td>DeepSeer-OCR (Tiny)</td></tr></table>

## 3. 带有标记 boxed 的识别
- `image_file = 'input/img/deepseekOCR.png'`

In [ ]:
from transformers import AutoModel, AutoTokenizer
import torch
import os
os.environ["CUDA_VISIBLE_DEVICES"] = '0'
model_name = 'deepseek-ai/DeepSeek-OCR'

tokenizer = AutoTokenizer.from_pretrained(model_name, trust_remote_code=True)
model = AutoModel.from_pretrained(model_name, _attn_implementation='flash_attention_2', trust_remote_code=True, use_safetensors=True)
model = model.eval().cuda().to(torch.bfloat16)

# prompt = "<image>\nFree OCR. "
prompt = "<image>\n<|grounding|>Convert the document to markdown. "

image_file = 'input/img/deepseekOCR.png'
output_path = 'img_out/Convert'

res = model.infer(tokenizer, prompt=prompt, image_file=image_file, output_path = output_path, base_size = 1024, image_size = 640, crop_mode=True, save_results = True, test_compress = True)


You are using a model of type deepseek_vl_v2 to instantiate a model of type DeepseekOCR. This is not supported for all configurations of models and can yield errors.
Some weights of DeepseekOCRForCausalLM were not initialized from the model checkpoint at deepseek-ai/DeepSeek-OCR and are newly initialized: ['model.vision_model.embeddings.position_ids']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:None for open-end generation.


BASE:  torch.Size([1, 256, 1280])
PATCHES:  torch.Size([8, 100, 1280])
<|ref|>image<|/ref|><|det|>[[40, 50, 485, 884]]<|/det|>
<|ref|>image_caption<|/ref|><|det|>[[110, 925, 420, 970]]<|/det|>
<center>(a) Compression on Fox benchmark </center>  

<|ref|>image<|/ref|><|det|>[[515, 30, 967, 884]]<|/det|>
<|ref|>image_caption<|/ref|><|det|>[[590, 925, 894, 970]]<|/det|>
<center>(b) Performance on Omnidocbench </center>
image size:  (2646, 1152)
valid image tokens:  911
output texts tokens (valid):  104
compression ratio:  0.11
===============save results:===============


other: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:00<00:00, 57456.22it/s]


In [ ]:
from transformers import AutoModel, AutoTokenizer
import torch
import os
os.environ["CUDA_VISIBLE_DEVICES"] = '0'
model_name = 'deepseek-ai/DeepSeek-OCR'

tokenizer = AutoTokenizer.from_pretrained(model_name, trust_remote_code=True)
model = AutoModel.from_pretrained(model_name, _attn_implementation='flash_attention_2', trust_remote_code=True, use_safetensors=True)
model = model.eval().cuda().to(torch.bfloat16)

# prompt = "<image>\nFree OCR. "
# prompt = "<image>\n<|grounding|>Convert the document to markdown. "
prompt = "figures in document: <image>\nParse the figure."

image_file = 'input/img/deepseekOCR.png'
output_path = 'img_out/Parse'

res = model.infer(tokenizer, prompt=prompt, image_file=image_file, output_path = output_path, base_size = 1024, image_size = 640, crop_mode=True, save_results = True, test_compress = True)


You are using a model of type deepseek_vl_v2 to instantiate a model of type DeepseekOCR. This is not supported for all configurations of models and can yield errors.
Some weights of DeepseekOCRForCausalLM were not initialized from the model checkpoint at deepseek-ai/DeepSeek-OCR and are newly initialized: ['model.vision_model.embeddings.position_ids']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/root/miniconda3/envs/deepseek-ocr/lib/python3.12/site-packages/transformers/generation/configuration_utils.py:590: UserWarning: `do_sample` is set to `False`. However, `temperature` is set to `0.0` -- this flag is only used in sample-based generation modes. You should set `do_sample=True` or unset `temperature`.
  warnings.warn(
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` t

BASE:  torch.Size([1, 256, 1280])
PATCHES:  torch.Size([8, 100, 1280])
<table><tr><td>Model</td><td>Text Tokens in Per Page (Ground-truth)</td><td>Overall Performance (Edit Distance)</td><td>Performance on Omnipodbench</td></tr><tr><td>DeepSeek-OCR (Gundam-M 200dpi)</td><td>600-700</td><td>10.5</td><td>High Accuracy ED < 0.25 (↑ better)</td></tr><tr><td>DeepSeek-OCR (Large)</td><td>700-800</td><td>11.8</td><td>High Accuracy ED < 0.25 (↑ better)</td></tr><tr><td>DeepSee-OCR (Small)</td><td>800-900</td><td>13.2</td><td>High Accuracy ED < 0.25 (↑ better)</td></tr><tr><td>Deep See-OCR (Tiny)</td><td>900-1000</td><td>8.5</td><td>High Accuracy ED < 0.25 (↑ better)</td></tr><tr><td rowspan="2">DeepSee-OCR (Tiny)</td><td>1000-1100</td><td>15.1</td><td>High Accuracy ED < 0.25 (↑ better)</td></tr><tr><td>1100-1200</td><td>79.3</td><td>High Accuracy ED < 0.25 (↑ better)</td></tr><tr><td>Deep Se-OCR (Tiny)</td><td>1200-1300</td><td>11.3</td><td>High Accuracy ED < 0.25 (↑ better)</td></tr><tr><td r

image: 0it [00:00, ?it/s]
other: 0it [00:00, ?it/s]


## 更复杂一点的图片（公式和sub-fig）
截取了deepseek math.pdf 里面的一页

![DeepSeek Math](./input/img/GRPO.png)

In [8]:
from transformers import AutoModel, AutoTokenizer
import torch
import os
os.environ["CUDA_VISIBLE_DEVICES"] = '0'
model_name = 'deepseek-ai/DeepSeek-OCR'

tokenizer = AutoTokenizer.from_pretrained(model_name, trust_remote_code=True)
model = AutoModel.from_pretrained(model_name, _attn_implementation='flash_attention_2', trust_remote_code=True, use_safetensors=True)
model = model.eval().cuda().to(torch.bfloat16)

# prompt = "<image>\nFree OCR. "
prompt = "<image>\n<|grounding|>Convert the document to markdown. "
# prompt = "figures in document: <image>\nParse the figure."

image_file = 'input/img/GRPO.png'
output_path = 'img_out/GRPO convert to md'

res = model.infer(tokenizer, prompt=prompt, image_file=image_file, output_path = output_path, base_size = 1024, image_size = 640, crop_mode=True, save_results = True, test_compress = True)


You are using a model of type deepseek_vl_v2 to instantiate a model of type DeepseekOCR. This is not supported for all configurations of models and can yield errors.
Some weights of DeepseekOCRForCausalLM were not initialized from the model checkpoint at deepseek-ai/DeepSeek-OCR and are newly initialized: ['model.vision_model.embeddings.position_ids']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:None for open-end generation.


BASE:  torch.Size([1, 256, 1280])
PATCHES:  torch.Size([6, 100, 1280])
<|ref|>image<|/ref|><|det|>[[50, 28, 940, 288]]<|/det|>
<|ref|>image_caption<|/ref|><|det|>[[21, 295, 988, 338]]<|/det|>
<center>Figure 4 | Demonstration of PPO and our GRPO. GRPO foregoes the value model, instead estimating the baseline from group scores, significantly reducing training resources. </center>  

<|ref|>text<|/ref|><|det|>[[21, 361, 988, 444]]<|/det|>
on the rewards \(\{r_{\geq t}\}\) and a learned value function \(V_{\psi}\) . Thus, in PPO, a value function needs to be trained alongside the policy model and to mitigate over- optimization of the reward model, the standard approach is to add a per- token KL penalty from a reference model in the reward at each token (Ouyang et al., 2022), i.e.,  

<|ref|>equation<|/ref|><|det|>[[325, 451, 988, 495]]<|/det|>
\[r_{t} = r_{\phi}(q,o_{\leq t}) - \beta \log \frac{\pi_{\theta}(o_{t}|q,o_{< t})}{\pi_{ref}(o_{t}|q,o_{< t})}, \quad (2)\]  

<|ref|>text<|/ref|><|

other: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 7/7 [00:00<00:00, 113798.95it/s]


![](img_out/GRPO%20convert%20to%20md/images/0.jpg)


<center>Figure 4 | Demonstration of PPO and our GRPO. GRPO foregoes the value model, instead estimating the baseline from group scores, significantly reducing training resources. </center>  


on the rewards \($\{r_{\geq t}\}$\) and a learned value function \($V_{\psi}$\) . Thus, in PPO, a value function needs to be trained alongside the policy model and to mitigate over- optimization of the reward model, the standard approach is to add a per- token KL penalty from a reference model in the reward at each token (Ouyang et al., 2022), i.e.,  


\[$$r_{t} = r_{\phi}(q,o_{\leq t}) - \beta \log \frac{\pi_{\theta}(o_{t}|q,o_{< t})}{\pi_{ref}(o_{t}|q,o_{< t})}, \quad (2)$$\]  


where \($r_{\phi}$\) is the reward model, \($\pi_{ref}$\) is the reference model, which is usually the initial SFT model, and \($\beta$\) is the coefficient of the KL penalty.  


As the value function employed in PPO is typically another model of comparable size as the policy model, it brings a substantial memory and computational burden. Additionally, during RL training, the value function is treated as a baseline in the calculation of the advantage for variance reduction. While in the LLM context, usually only the last token is assigned a reward score by the reward model, which may complicate the training of a value function that is accurate at each token. To address this, as shown in Figure 4, we propose Group Relative Policy Optimization (GRPO), which obviates the need for additional value function approximation as in PPO, and instead uses the average reward of multiple sampled outputs, produced in response to the same question, as the baseline. More specifically, for each question \($q$\) , GRPO samples a group of outputs \($\{o_{1}, o_{2}, \dots , o_{G}\}$\) from the old policy \(\pi_{\theta_{old}}\) and then optimizes the policy model by maximizing the following objective:  


\[$$\begin{array}{l}{\mathcal{I}_{GRPO}(\theta) = \mathbb{E}[q\sim P(Q),\{o_{i}\}_{i = 1}^{G}\sim \pi_{\theta_{old}}(O|q)]}\\ {\frac{1}{G}\sum_{i = 1}^{G}\frac{1}{|o_{i}|}\sum_{t = 1}^{|o_{i}|}\Big\{\min \Big[\frac{\pi_{\theta}(o_{i,t}|q,o_{i,< t})}{\pi_{\theta_{old}}(o_{i,t}|q,o_{i,< t})}\hat{A}_{i,t},\mathrm{clip}\Big(\frac{\pi_{\theta}(o_{i,t}|q,o_{i,< t},}{\pi_{\theta_{old}}(o_{i,t}|q,o_{i,< \epsilon})},1 - \epsilon ,1 + \epsilon \Big)\hat{A}_{i,t}\Big] - \beta \mathbb{D}_{KL}\big[\pi_{\theta}||\pi_{ref}\big]\Big\} ,} \end{array} \quad (3)$$\]  


where \($\epsilon$\) and \($\beta$\) are hyper- parameters, and \($\hat{A}_{i,t}$\) is the advantage calculated based on relative rewards of the outputs inside each group only, which will be detailed in the following subsections. The group relative way that GRPO leverages to calculate the advantages, aligns well with the comparative nature of rewards models, as reward models are typically trained on datasets of comparisons between outputs on the same question. Also note that, instead of adding KL penalty in the reward, GRPO regularizes by directly adding the KL divergence between the trained policy and the reference policy to the loss, avoiding complicating the calculation of \($\hat{A}_{i,t}$\) .